util/src/main/java/io/kubernetes/client/Exec.java

```diff
@@ -191,6 +199,135 @@ public Process exec(
+   */
+  public Future<Integer> exec(
+      String namespace,
+      String podName,
+      Consumer<IOTrio> onOpen,
+      BiConsumer<Integer, IOTrio> onClosed,
+      BiConsumer<Throwable, IOTrio> onError,
+      Long timeoutMs,
+      boolean tty,
+      String... command)
+      throws IOException {
+    CompletableFuture<Integer> future = new CompletableFuture<>();
+    IOTrio io = new IOTrio();
+    String cmdStr = Arrays.toString(command);
+    try {
+      Process process =
+          newExecutionBuilder(namespace, podName, command)
+              .setStdin(true)
+              .setStdout(true)
+              .setStderr(true)
+              .setTty(tty)
+              .execute();
+      io.onClose(
+          (code, timeout) -> {
+            process.destroy();
+            waitForProcessToExit(process, timeout, cmdStr, err -> onError.accept(err, io));
+            // processWaitingThread will handle the rest
+          });
+      io.setStdin(process.getOutputStream());
+      io.setStdout(process.getInputStream());
+      io.setStderr(process.getErrorStream());
+      Thread processWaitingThread =
+          new Thread(
+              () -> {
+                Supplier<Integer> returnCode = process::exitValue;
+                try {
+                  log.debug("Waiting for process to close in {} ms: {}", timeoutMs, cmdStr);
+                  boolean beforeTimout =
+                      waitForProcessToExit(
+                          process, timeoutMs, cmdStr, err -> onError.accept(err, io));
+                  if (!beforeTimout) {
+                    returnCode = () -> Integer.MAX_VALUE;
+                  }
+                } catch (Exception e) {
+                  onError.accept(e, io);
+                }
+                log.debug("process.onExit({}): {}", returnCode.get(), cmdStr);
+                future.complete(returnCode.get());
+                onClosed.accept(returnCode.get(), io);
+              });
+      processWaitingThread.setName("Process-Waiting-Thread-" + command[0] + "-" + podName);
+      processWaitingThread.start();
+      onOpen.accept(io);
+    } catch (ApiException e) {
+      onError.accept(e, io);
+      future.completeExceptionally(e);
+    }
+    return future;
+  }
+
+  private boolean waitForProcessToExit(
+      Process process, Long timeoutMs, String cmdStr, Consumer<Exception> onError) {
+    boolean beforeTimeout = true;
+    if (timeoutMs != null && timeoutMs >= 0) {
+      boolean exited = false;
+      try {
+        exited = process.waitFor(timeoutMs, TimeUnit.MILLISECONDS);
+      } catch (InterruptedException e) {
+        onError.accept(e);
+      }
+      log.debug("Process closed={}: {}", exited, cmdStr);
+      if (!exited && process.isAlive()) {
+        beforeTimeout = false;
+        log.warn("Process timed out after {} ms. Shutting down: {}", timeoutMs, cmdStr);
+        process.destroy();
+        try {
+          exited = process.waitFor(graceTimeoutMs, TimeUnit.MILLISECONDS);
```

Comment:
It's a little weird to wait after the user specified timeout. I think we should delete all of the grace period logic. The user should just pass in whatever timeout they want.

Reply:
Have you ever closely monitored a Tomcat process shutting down? :) It takes a few seconds to clear all the pools, close all the connections, commit all the remaining transactions after the JVM receives a signal #15. Similar delays are observed in multi-classloader applications (e.g. osgi-based). Shutting down the application gracefully might take a tad more than just a heartbeat.

In line 303 I'm (hopefully) sending a sigterm, so the application starts its shutdown sequence. Some applications might take considerable amounts of time to shutdown (e.g. a JVM with a debugger attached), thus I cannot guarantee the process will be stopped immediately after the timeout in line #295. Telling the caller the process has stopped could lead to resources' exhaustion in some scenarios (i.e. BUGS). So I'm giving the process a chance to finish its business itself in 5 seconds (by default). If it fails to do so (could be many reasons), then I'm using my last card - sending (hopefully) a sigkill.

Sending a sigkill must always be the last resort.

I'd really like to keep the grace period in the code. I can hardcode the grace period if you'd like me to (could be extracted later on if required), but I'd sleep better knowing I've pushed the correct process termination procedure to a public repo :)

Reply:
But this isn't a real process, it's a remote process running on a different service in a container on the cluster, so I don't think that stuff applies here.

Reply:
That's the obvious consequence of Liskov's principle violation :) Frankly, I don't know what does this client does when I ask it to terminate. So which one does reliably destroy the connection and remote process? destroy() or destroyForcibly()? I'll use the one you name and remove the grace period then.

Reply:
`destroy()` does the right thing here (it's not 100% 'reliable' in the sense that nothing in distributed systems is 100% "reliable"), but `destroy` closes the web socket connection which should have the net effect of stopping the process on the remote server (eventually), but from the client's perspective, it's terminated once the socket is closed.

Code is here:
https://github.com/kubernetes-client/java/blob/master/util/src/main/java/io/kubernetes/client/Exec.java#L476